src/stm32/stm32h7_adc.c

```diff
@@ -303,9 +299,11 @@ gpio_adc_cancel_sample(struct gpio_adc g)
 {
     ADC_TypeDef *adc = g.adc;
     irqstatus_t flag = irq_save();
-    // ADSTART is not as long true as SR_STRT on stm32f4
-    if ((adc->CR & ADC_CR_ADSTART || adc->ISR & ADC_ISR_EOC)
-        && adc->SQR1 == (g.chan << ADC_SQR1_SQ1_Pos))
-        gpio_adc_read(g);
+    if (adc->SQR1 == (g.chan << ADC_SQR1_SQ1_Pos)) {
+        if (adc->CR & ADC_CR_ADSTART)
+            adc->CR |= ADC_CR_ADSTP;
```

Comment:
It makes me a little nervous that there may be a race with this read modify write compared to the f0 version which is only a write. With very unfortunate timing, ADSTART may clear between the read and the write. 

Based on this excerpt from the L412 manual, it suggests ADSTP might be blocked from occurring (which is okay because it has already stopped), but it might be able to set the ADSTART bit.

"The software is allowed to write the ADSTP or JADSTP control bits of the ADC_CR register
only if the ADC is enabled, possibly converting, and if there is no pending request to disable
the ADC (ADSTART or JADSTART must be equal to 1 and ADDIS to 0)."

Reply:
Good catch!  I modified this PR to try to avoid that race.

In general, I think stm32h7_adc.c should be reworked to avoid read-modify-write on all `ADC->CR` accesses.  That register has side-effects so using read-modify-write is just asking for subtle race conditions.  However, I don't want to conflict with your stm32g4 PR nor the stm32h723 PR.  So, something to look at after those PRs are merged.

-Kevin